### PR TITLE
feat(api): add enpoint for club registration

### LINKED ISSUE #202: [API] Endpoint for club registration

### PR DESCRIPTION
**Pull request by PranshuNayak**
Other things are the same. I have made changes in `api/clubs/serializers.py` and `api/clubs/tests.py` with respect to the following requested changes
review_request
.

### CODE REVIEW

File: api/api/clubs/tests.py [hunk 1]
```diff
@@ -191,3 +192,178 @@ def testUpdateClubs_positionHolderRequest_updatesSuccessful(position):
...
+        'fee_submitted': True,
+        'remark': 'Late fees',
+        'status': 'Approved',
+    })
+
+    # then
+    updated_registration_request = ClubRegistrationRequest.objects.get(
+        pk=test_registration_request.id)
+    assert response.status_code == status.HTTP_200_OK
+    assert response.data, {'id': test_registration_request.id,
```
Comment (Samy-33) [inline api/api/clubs/tests.py:319]: Shouldn't this be ==? instead of the comma?
other places too.

File: api/api/clubs/tests.py [hunk 2]
```diff
@@ -191,3 +192,178 @@ def testUpdateClubs_positionHolderRequest_updatesSuccessful(position):
...
+                           'updated_by': user.id}
+    assert updated_registration_request.fee_submitted
+    assert updated_registration_request.remark == 'Late fees'
+    assert updated_registration_request.status == 'Approved'
+    assert updated_registration_request.updated_by == user
+
+
+@pytest.mark.django_db
+@pytest.mark.parametrize('position', ['Coordinator', 'Co-Coordinator', 'Core member'])
+def testUpdateClubRegistrationRequest_coreMemberRequest_pass(client, position, test_user,
```
Comment (Samy-33) [inline api/api/clubs/tests.py:335]: instead of pass, let's add what happens i.e. updatesTheRequest

File: api/api/clubs/urls.py [hunk 3]
```diff
@@ -4,5 +4,7 @@
 
 urlpatterns = [
     path('', views.ListClubsView.as_view()),
-    path('<int:pk>/', views.UpdateClubsView.as_view())
+    path('<int:pk>/', views.UpdateClubsView.as_view()),
+    path('<int:pk>/register/', views.CreateClubRegistrationRequestView.as_view()),
+    path('registration/<int:pk>/', views.UpdateClubRegistrationRequestView.as_view())
```
Comment (Samy-33) [inline api/api/clubs/urls.py:9]: url should be `<int:pk>/registration-requests/<int:pk`

File: api/api/clubs/urls.py [hunk 4]
```diff
@@ -4,5 +4,7 @@
 
 urlpatterns = [
     path('', views.ListClubsView.as_view()),
-    path('<int:pk>/', views.UpdateClubsView.as_view())
+    path('<int:pk>/', views.UpdateClubsView.as_view()),
+    path('<int:pk>/register/', views.CreateClubRegistrationRequestView.as_view()),
```
Comment (Samy-33) [inline api/api/clubs/urls.py:8]: here url should be `<int:pk>/registration-requests/`

This document should be helpful in understanding what the urls should be: https://restfulapi.net/resource-naming/

**Review by Samy-33 — CHANGES_REQUESTED** [review]
Reviewed partially, will review again.

File: api/api/clubs/tests.py (hunk 1 above)
Comment (PranshuNayak) [reply api/api/clubs/tests.py:319]: Typo

File: api/api/clubs/urls.py (hunk 3 above)
Comment (PranshuNayak) [reply api/api/clubs/urls.py:9]: Again don't understand the need for `<int:pk>/registration-request<...>` . I suppose we only need registration id to update a particular request .

Comment (PranshuNayak) [reply api/api/clubs/urls.py:9]: This change will only affect the URL ? coz current implementation deals only with `registration id`

Comment (Samy-33) [reply api/api/clubs/urls.py:9]: Registration request is a child resource of club. So in the implementation find should be called with club id and registration request id

**Comment by PranshuNayak** [comment]
Any idea ?
Using `datetime.now(timezone.utc)` in RH
erro

**Comment by PranshuNayak** [comment]
Caught in this error again 

Here is defined url in `api/clubs/urls.py`
` path('<int:pk_club>/registration-requests/<int:pk>/', views.UpdateClubRegistrationRequestView.as_view())`
erro

**Comment by PranshuNayak** [comment]
Caught in this error again . Don't know why is does not recognize the url . Working fine in browser.
urlPatterns : `path('<int:pk_1>/registration-requests/<int:pk_2>/', views.UpdateClubRegistrationRequestView.as_view())`
error

**Comment by PranshuNayak** [comment]
@Samy-33 ready for review

File: api/api/clubs/urls.py [hunk 5]
```diff
@@ -4,5 +4,8 @@
 
 urlpatterns = [
     path('', views.ListClubsView.as_view()),
-    path('<int:pk>/', views.UpdateClubsView.as_view())
+    path('<int:pk>/', views.UpdateClubsView.as_view()),
+    path('<int:pk>/registration-requests/', views.CreateClubRegistrationRequestView.as_view()),
+    path('<int:pk_1>/registration-requests/<int:pk_2>/',
```
Comment (Samy-33) [inline api/api/clubs/urls.py:9]: let's give pk's more readable names:
for example: pk_1 can be -> club_id, pk_2 can be registration_id

File: api/api/clubs/views.py [hunk 6]
```diff
@@ -14,3 +15,22 @@ class UpdateClubsView(UpdateAPIView):
...
+    serializer_class = ClubUpdateRegistrationRequestSerializer
+    # pk_2 is registration request id that is requested to be updated
+    lookup_url_kwarg = "pk_2"
+
+    def get_queryset(self):
+        # pk_1 is the club id wrt to given registration request
+        queryset = ClubRegistrationRequest.objects.filter(club=self.kwargs['pk_1'])
+        if queryset:
+            return queryset
+        return ClubRegistrationRequest.objects.none()
```
Comment (Samy-33) [inline api/api/clubs/views.py:36]: Does these lines achieve anything?

We could simply return `ClubRegistrationRequest.objects.filter(club=self.kwargs['pk_1'])`

**Review by Samy-33 — CHANGES_REQUESTED** [review]

**Review by Samy-33 — DISMISSED** [review]

**Comment by PranshuNayak** [comment]
Changed permissions to allow only authorisied roles to access the update registration endpoint.

File: api/api/clubs/permissions.py [hunk 7]
```diff
@@ -16,6 +16,13 @@ def has_object_permission(self, request, view, obj):
 class IsCoreMemberOrAdmin(permissions.BasePermission):
     """Custom permission to allow only core members of respective club and admins"""
 
+    def has_permission(self, request, view):
+        user_roles = request.user.roles.filter(Q(name='Coordinator') | Q(
+            name='Co-Coordinator') | Q(name='Core member'))
```
Comment (Samy-33) [inline api/api/clubs/permissions.py:21]: Can we update these strings to variables from model here and other places in this file?

**Comment by PranshuNayak** [comment]
Pending for review

**Review by Samy-33 — APPROVED** [review]

**Pull request merged**